grails-app/controllers/org/pih/warehouse/shipping/ShipmentController.groovy

```diff
@@ -888,13 +888,13 @@ class ShipmentController {
 
 
     def saveComment() {
-        def shipmentInstance = Shipment.get(params.shipmentId)
-        def recipient = (params.recipientId) ? User.get(params.recipientId) : null
-        def comment = new Comment(comment: params.comment, sender: session.user, recipient: recipient)
-        if (shipmentInstance) {
-            shipmentInstance.addToComments(comment).save(flush: true)
-            flash.message = "${warehouse.message(code: 'shipping.addedCommentToShipment.message', args: [params.comment, shipmentInstance.name])}"
+        Shipment shipmentInstance = Shipment.get(params.shipmentId)
+        if (!shipmentInstance) {
+            throw new Exception("Could not locate shipment with ID " + params.shipmentId)
         }
+        User recipient = (params.recipientId) ? User.get(params.recipientId) : null
+        shipmentService.addShipmentComment(shipmentInstance, params.comment,  session.user, recipient)
+        flash.message = "${warehouse.message(code: 'shipping.addedCommentToShipment.message', args: [params.comment, shipmentInstance.name])}"
```

Comment:
the problem with that is since the `ShipmentController` is `@Transactional`, you start and end the transaction boundary in the controller, hence if anything goes wrong while persisting the comment, you will still get the success message in the view, because the exception will happen at the end of the session, not while executing the `896`th line.

Reply:
I know we may have encountered issues like this in the past, but let's confirm this the actual behavior.